### PR TITLE
fix: Fix image flickering

### LINKED ISSUE #4473: [WASM] BitmapIcon blinking invalid image on hover
## Current behavior

blinking2

## Expected behavior

Should not blink.

## How to reproduce it (as minimally and precisely as possible)

```
<Button
    Grid.RowSpan="2"
    Grid.ColumnSpan="2"
    HorizontalAlignment="Right"
    VerticalAlignment="Bottom"
    Background="Transparent"
    BorderBrush="Transparent"
    Command="{x:Bind ViewModel.OpenSettingsCommand}">
    <BitmapIcon
        Width="48"
        Height="48"
        ShowAsMonochrome="false"
        UriSource="/Assets/Settings.png" />
</Button>
```

## Workaround

N/A

## Environment

Nuget Package:

- [x] Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia
- [ ] Uno.WinUI / Uno.WinUI.WebAssembly / Uno.WinUI.Skia
- [ ] Uno.SourceGenerationTasks
- [ ] Uno.UI.RemoteControl / Uno.WinUI.RemoteControl
- [ ] Other: 

Nuget Package Version(s): 3.2.0-dev.401

Affected platform(s):

- [ ] iOS
- [ ] Android
- [x] WebAssembly
- [ ] WebAssembly renderers for Xamarin.Forms
- [ ] macOS
- [ ] Skia
  - [ ] WPF
  - [ ] GTK (Linux)
  - [ ] Tizen
- [ ] Windows
- [ ] Build tasks
- [ ] Solution Templates

IDE:

- [ ] Visual Studio 2017 (version: )
- [x] Visual Studio 2019 (version: )
- [ ] Visual Studio for Mac (version: )
- [ ] Rider Windows (version: )
- [ ] Rider macOS (version: )
- [ ] Visual Studio Code (version: )

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): closes #4473

## PR Type

What kind of change does this PR introduce?

- Bugfix

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI/UI/Xaml/Controls/BitmapIcon.cs [hunk 1]
```diff
@@ -67,14 +67,6 @@ private void RefreshImage()
...
 #if !NET461
 			UpdateImageMonochromeColor();
-
-			if (UriSource != null)
-			{
-				// Force a reload
-				var tmp = UriSource;
-				UriSource = null;
-				UriSource = tmp;
-			}
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/BitmapIcon.cs:77]: This part is needed for reloading on other targets. We may need to adjust this specifically for webassembly.

**Review by jeromelaban — CHANGES_REQUESTED** [review]

**Comment by Youssef1313** [comment]
@jeromelaban Is anything still missing here?

**Comment by jeromelaban** [comment]
@Youssef1313 This is likely to not be enough. In this case, if we have a change in "ShowAsMonochrome" the image won't be refreshed properly. It's likely that the image refreshes itself because the foreground color has changed, something we do not need to do if we don't use the `ShowAsMonochrome` property.

Changing this would still provoque a flicker when using the `ShowAsMonochrome`. We should add additional UI tests to validate that changes to the `Foreground` and `ShowAsMonochrome` are actually updating the image accordingly.

**Comment by Youssef1313** [comment]
There are already tests for monochrome images: https://github.com/unoplatform/uno/blob/master/src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/BitmapIconTests/BitmapIcon_Generic.xaml

Wondering how they're passing if my change is supposed to break them.

**Comment by Youssef1313** [comment]
I also just validated the behavior of [the existing test](https://github.com/unoplatform/uno/blob/master/src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/BitmapIconTests/BitmapIcon_Generic.xaml) with and without my change. Changes to Foreground update the image correctly in both cases.

**Comment by jeromelaban** [comment]
Good, then I wonder what's the source of the refresh then, if it's not the foreground change. Do you know which part does this?

**Comment by Youssef1313** [comment]
I don't :(

btw, just a clarification to my previous comment: the flickering I'm referring to is when clicking the "change color" button in the monochrome test, while the one in the issue worked on for this PR happens when hovering.

I wonder if it's okay to keep the current change as is until we get any feedback that it broke something? or if someone knows a particular issue can be caused by this change?

**Comment by Youssef1313** [comment]
Would someone be able to help here? cc @jeromelaban @carldebilly

**Comment by MartinZikmund** [comment]
As long as the monochrome foreground color change still updates properly, it should be ok.

**Review by MartinZikmund — APPROVED** [review]

**Comment by jeromelaban** [comment]
I've been looking again at this PR, and I don't see how this can work. `_image.MonochromeColor` is only read when the Uri changes, which means that removing cycling through null should also break the Foreground color change. Let's restart a build to take a look at the sample mentioned above, it's likely that we're missing something.

**Comment by jeromelaban** [comment]
I just confirmed, the color [of this test](https://github.com/unoplatform/uno/blob/master/src/SamplesApp/UITests.Shared/Windows_UI_Xaml_Controls/BitmapIconTests/BitmapIcon_Generic.xaml.cs) does not change with the reload commented out.

The original issue was about the fact that the icon was reloaded. We probably should be avoiding the reload if the foreground changes, but icon is not monochrome.

**Review by Youssef1313 — COMMENTED** [review]
Thanks @jeromelaban. Can you take a look at the new approach? I'm now forcing a reload without temporarily setting `UriSource` to null.

**Comment by jeromelaban** [comment]
> Thanks @jeromelaban. Can you take a look at the new approach? I'm now forcing a reload without temporarily setting `UriSource` to null.

This looks a lot better, thanks :) This PR has shown something though, which is that the validation of the color changing is not done properly in the UI tests. This test:

https://github.com/unoplatform/uno/blob/ece6481aec938528b3420fe6e9322584436787c3/src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/BitmapIconTests/UnoSamples_Tests.BitmapIcon.cs

Should also validate that the color is actually changed visually (searching for the new color in the new `Image` should do).

**Review by Youssef1313 — COMMENTED** [review]
I'm seeing the following exception when running UI Tests:

```csharp
    OneTimeSetUp: System.TypeInitializationException : The type initializer for 'SamplesApp.UITests.SampleControlUITestBase' threw an exception.
      ----> System.Exception : Failed to execute: C:\Program Files (x86)\Android\android-sdk\platform-tools\adb.exe -s emulator-5554 forward tcp:56863 tcp:56863 - exit code: 1
    adb.exe: error: cannot bind listener: cannot bind to 127.0.0.1:56863: An attempt was made to access a socket in a way forbidden by its access permissions. (10013)

```

Have been encountering that since yesterday. @jeromelaban Any idea about that?

**Comment by MartinZikmund** [comment]
@Youssef1313 did you try rebooting PC 🙂 ?

**Comment by Youssef1313** [comment]
Thanks @jeromelaban @MartinZikmund! Fixed :)

File: src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/BitmapIconTests/UnoSamples_Tests.BitmapIcon.cs [hunk 2]
```diff
@@ -14,28 +15,49 @@ namespace SamplesApp.UITests.Windows_UI_Xaml_Controls.BitmapIconTests
...
+					var pixel = bitmap.GetPixel(x, y);
+					if ((pixel.A, pixel.R, pixel.G, pixel.B) == (color.A, color.R, color.G, color.B))
+					{
+						return true;
+					}
+				}
+			}
+
+			return false;
+		}
```
Comment (jeromelaban) [inline src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/BitmapIconTests/UnoSamples_Tests.BitmapIcon.cs:33]: ImageHelper likely contains this method.

Comment (Youssef1313) [reply src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/BitmapIconTests/UnoSamples_Tests.BitmapIcon.cs:33]: @jeromelaban I couldn't find it anywhere.

Comment (jeromelaban) [reply src/SamplesApp/SamplesApp.UITests/Windows_UI_Xaml_Controls/BitmapIconTests/UnoSamples_Tests.BitmapIcon.cs:33]: Then let's add it in there then, if nothing matches the behavior you're adding: https://github.com/unoplatform/uno/blob/master/src/SamplesApp/SamplesApp.UITests/TestFramework/ImageAssert.cs

**Review by jeromelaban — APPROVED** [review]

**Comment by Youssef1313** [comment]
@jeromelaban Anything left here?

**Pull request merged**